### PR TITLE
CDAP-12560 Emit Task-level metrics from Mapper/Reducer Tasks

### PR DESCRIPTION
**Pull request by gokulavasan**
Instead of emitting it in the MapReduce driver

JIRA: https://issues.cask.co/browse/CDAP-12560
Build: https://builds.cask.co/browse/CDAP-RUT1402

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 1]
```diff
@@ -167,6 +167,7 @@
 
   private Job job;
   private Runnable cleanupTask;
+  private int metricsEmitIntervalInSecs;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: Why this has to be a field? I think you can turn the whole logic into a private method `getMetricsReportIntervalMillis(CConfiguration cConf, Map<String, String> runtimeArgs)`. It is ok to call this method multiple times.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/TaskMetricsWriter.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * the metrics system.
+ */
+public abstract class TaskMetricsWriter {
+  public static final String PROPERTY_NAME = "c.mr.task.metrics.emit.interval.in.secs";
+  private final TaskInputOutputContext taskContext;
+  private final MetricsContext metricsContext;
+
+  TaskMetricsWriter(AbstractContext context, TaskInputOutputContext taskContext) {
+    this.taskContext = taskContext;
+    this.metricsContext = context.getProgramMetrics();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/TaskMetricsWriter.java]: If you only need the `MetricsContext`, have it pass from the constructor directly. This make the class more generic and easier to use/test/mock.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/TaskMetricsWriter.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.hadoop.mapreduce.TaskInputOutputContext;
+
+import java.util.Map;
+
+/**
+ * Base class for gathering statistics from a running map/reduce task through its counters and for writing the data to
+ * the metrics system.
+ */
+public abstract class TaskMetricsWriter {
+  public static final String PROPERTY_NAME = "c.mr.task.metrics.emit.interval.in.secs";
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/TaskMetricsWriter.java]: What is this for?? Can't we just take cConf from the constructor?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/ReducerWrapper.java [hunk 4]
```diff
@@ -69,6 +76,15 @@ public void run(Context context) throws IOException, InterruptedException {
     WrappedReducer.Context flushingContext = createAutoFlushingContext(context, basicMapReduceContext);
     basicMapReduceContext.setHadoopContext(flushingContext);
 
+    int metricsReportInterval = context.getConfiguration().getInt(TaskMetricsWriter.PROPERTY_NAME, 60);
+    final ReduceTaskMetricsWriter reduceTaskMetricsWriter = new ReduceTaskMetricsWriter(basicMapReduceContext, context);
+    metricsEmitterService.scheduleAtFixedRate(new Runnable() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/ReducerWrapper.java]: It's better not to start a new thread as I am not sure if the task context is thread safe or not. It's better do it in the `WrappedMapper.Context` and `WrappedReducer.Context` when data are read/write.

**Comment by gokulavasan** [comment]
Thanks for the review @chtyim. Addressed comments. Please take a look again.

**Comment by chtyim** [comment]
LGTM.

**Pull request merged**